objects/all.go

```diff
@@ -17,25 +17,162 @@ limitations under the License.
 package objects
 
 import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/kustomize/v3/k8sdeps/kunstruct"
+	"sigs.k8s.io/kustomize/v3/k8sdeps/transformer"
+	"sigs.k8s.io/kustomize/v3/k8sdeps/validator"
+	"sigs.k8s.io/kustomize/v3/pkg/fs"
+	"sigs.k8s.io/kustomize/v3/pkg/ifc"
+	"sigs.k8s.io/kustomize/v3/pkg/loader"
+	"sigs.k8s.io/kustomize/v3/pkg/resmap"
+	"sigs.k8s.io/kustomize/v3/pkg/resource"
+	"sigs.k8s.io/kustomize/v3/pkg/target"
+	"sigs.k8s.io/kustomize/v3/pkg/transformers"
 )
 
-// GetManegementCluster returns all the objects needed to create a working CAPD management cluster
-func GetManegementCluster(version, capiImage, capdImage string) ([]runtime.Object, error) {
-	capiObjects, err := GetCAPI(version, capiImage)
+// Provider is any of the required capi-providers (bootstrap, infrastructure, capi (core))
+type Provider struct {
+	// Name is the name of the provider, the Github project name
+	Name string
+	// Manager image the name of the manager image for this provider
+	ManagerImage string
+
+	// Version is a git ref that you want to use
+	Version string
+
+	// Organization is the github organization
+	Organization string
+
+	// The local or remote path to a CRD
+	CRDPath string
+
+	// CustomImage will override whatever image this provier uses for the manager
+	CustomImage string
+	// The kubernetes Kind of the manager (Deployment, StatefulSet, etc)
+	ManagerKind string
+}
+
+// GetCRDPath generates a remote reference for kustomize or returns the provided value
+func (p *Provider) GetCRDPath() string {
+	if p.CRDPath != "" {
+		return p.CRDPath
+	}
+	return fmt.Sprintf("https://github.com/%s/%s/config/default?ref=%s", p.Organization, p.Name, p.Version)
+}
+
+// OverrideImage is a MutateFunction found in Kustomize's transform.
+func (p *Provider) OverrideImage(item interface{}) (interface{}, error) {
+	// Oh yes, we are sure this is a string
+	s := item.(string)
+	// IMAGE_URL is the default provided by kubebuilder without ever generating a real patch
+	if strings.Contains(s, "manager") || s == "IMAGE_URL" || strings.Contains(s, "cluster-api-controller") {
```

Comment:
this needs to be smarter. 

Right now, if you look at line 122 and read Kustomize's code, you will see that we can only select every element in a list, not just a single particular container. Because of that we have to figure out which image value to actually update.

Reply:
Hi @chuckha,

Did you see the work I posted about last week regarding CAPV's `manifests` command and virtual patches for `kustomize`?

I recommend taking a look at the [documentation](https://github.com/akutz/cluster-api-provider-vsphere/tree/feature/go-manifests/cmd/manifests#Virtual-patches) I wrote about virtual patches. It may solve things for you in a manner that circumvents some of the issues you're encountering.

Reply:
I haven't yet, I'll add that as an issue to investigate later!